### PR TITLE
Allow annotation images to be updated

### LINKED ISSUE #2517: Modifying annotation should not require removing it and adding it again (iOS)
Currently, the only way to get the map to show changes in an annotation (especially its image) is is to remove it and add it again i.e. 

```
[self.mapView removeAnnotation:annotation];
[self.mapView addAnnotation:annotation];
```

This causes flickering (see #1688) but also has the side effect of de-selecting an annotation that was previously selected by the user. There is a workaround for this:

```
BOOL annotationWasSelected = [self.mapView.selectedAnnotations containsObject:annotation];
[self.mapView removeAnnotation:annotation];
[self.mapView addAnnotation:annotation];
if(annotationWasSelected) [self.mapView selectAnnotation:annotation animated:false];
```

But ideally it should be possible to modify an annotation is a simpler way.

### PR DESCRIPTION
**Pull request by RomainQuidet**
Add possibility to delete / update a sprite in order to update an AnnotationImage

#2210
Fixes #2517

### CODE REVIEW

**Comment by incanus** [comment]
Thanks for this @RomainQuidet. May take a bit, but we will try to review this soon. It looks pretty clean. 

/cc @kkaefer @jfirebaugh @1ec5

File: include/mbgl/ios/MGLAnnotationImage.h [hunk 1]
```diff
@@ -13,12 +13,12 @@ NS_ASSUME_NONNULL_BEGIN
 *   @param image The image to be displayed for the annotation.
 *   @param reuseIdentifier The string that identifies that this annotation image is reusable.
 *   @return The initialized annotation image object or `nil` if there was a problem initializing the object. */
-+ (instancetype)annotationImageWithImage:(UIImage *)image reuseIdentifier:(NSString *)reuseIdentifier;
++ (instancetype)annotationImageWithImage:(UIImage *)image reuseIdentifier:(nullable NSString *)reuseIdentifier;
```
Comment (1ec5) [inline include/mbgl/ios/MGLAnnotationImage.h]: MGLMapView currently assumes that the reuse identifier is non-null. Is there a need for it to be nullable?

File: src/mbgl/sprite/sprite_atlas.cpp [hunk 2]
```diff
@@ -187,8 +187,15 @@ void SpriteAtlas::updateDirty() {
             // The two names match;
             Holder& holder = imageIterator->second;
             holder.texture = spriteIterator->second;
-            copy(holder, imageIterator->first.second);
-
+            if (holder.texture != nullptr)
+            {
```
Comment (1ec5) [inline src/mbgl/sprite/sprite_atlas.cpp]: Nit: The C++ side of this project places braces on the same line, in contrast to the prevailing Objective-C style around here. (I’d be in favor of changing the Objective-C style to same-line braces for consistency.)

File: include/mbgl/ios/MGLAnnotationImage.h (hunk 1 above)
Comment (RomainQuidet) [inline include/mbgl/ios/MGLAnnotationImage.h]: Hi,
That was a reminder to make it nullable as in the official MapKit. This identifier is used to recycle the instance for performance reason, it should not be used internally like today to reference the image view in a dictionary or even deeper with the sprites

**Comment by RomainQuidet** [comment]
@1ec5 : do you need me to update the pull request or will you merge it ?

**Comment by 1ec5** [comment]
Since we use fast-forward merging, it'd be cleaner for you to rebase and resolve the conflicts. I'd be happy to do it but I'd have to open a new PR and close this one.

**Comment by RomainQuidet** [comment]
@1ec5 : branch rebased and corrections done according to your code review

**Comment by 1ec5** [comment]
Merged in 9d0ab5549ceb9da57293cf5259a4a00eb4383260...82a55755135c45f3a5ea3dab5a8acd9ee1034ee2. The test failure above is a random KIF failure, and Bitrise won’t let me rebuild the fork branch.

Thanks @RomainQuidet!

**Comment by alexagat** [comment]
When I update the `image` property on a `MGLAnnotationImage`, for example in the `didSelectAnnotation` function:
```
func mapView(mapView: MGLMapView, didSelectAnnotation annotation: MGLAnnotation) {
    annotationsTitlesAndImages[annotation.title!!]!.image = UIImage(named: "map-pin-active")!
}
```
...I see the below block of code in mapbox-gl framework run, but the annotation image does not update:

```
- (void)setImage:(UIImage *)image {
    _image = image;
    [self.delegate annotationImageNeedsRedisplay:self];
}
```

Is this still an issue?

**Comment by 1ec5** [comment]
@alexagat, can you open a new issue about this? It sounds like a regression if that’s the case.

**Pull request closed without merging**